### ISSUE TITLE
bff deduplication removes >90% data with NaiveBoth remove type

### ISSUE THREAD

**Issue by XirenZhou** [mlfoundations/dclm#83]
This issue has also been reported by others: https://github.com/mlfoundations/dclm/issues/71
Even you set `expected-ngram-count` to a large enough number, you would still end up with the majority of the data being removed. So this shouldn't be the major cause.

I guess I've found out the real root cause here: **the [dedup script](https://github.com/mlfoundations/dclm/blob/main/dedup/bff/src/main.rs#L726C13-L736C14) mistakenly updates the bloom filter twice (both after document-level dedup and after paragraph-level dedup)**. Therefore, even if a document survives a document-level removal, all of its ngrams have unfortunately already been inserted into the bloom filter (see https://github.com/mlfoundations/dclm/blob/main/dedup/bff/src/main.rs#L869); as a result, when the same document undergoes the paragraph-level removal, almost all the 13-word ngrams are already in the bloom filter and therefore get removed. (The reason that removal rate != 100% is that, 1. there are texts with num_tokens < min_ngram_size(i.e. 13); 2. paragraph-level dedup splits the document by "\n" which may produces paragraphs with small num_tokens whereas document-level dedup treats the document as a whole.

I'm wondering why you are doing the NaiveBoth type dedup [this way](https://github.com/mlfoundations/dclm/blob/main/dedup/bff/src/main.rs#L726C13-L736C14):
![image](https://github.com/user-attachments/assets/f689cf81-788a-456f-bcba-b795de507cf2)
instead of just doing
```rust
RemoveType::NaiveBoth => {
    process_line(&line, &bloom_filter, min_ngram_size, max_ngram_size,
        &RemoveType::Both, filtering_threshold, no_update_bloom_filter, annotate)       
}
```
The latter results in a consistent removal rate with the original paper.

**Comment by revbucket** [comment]
Oh...um, yep, that's a good point!

So what we ended up doing in the actual paper (and what we used to generate our datasets) maps directly to the method `--remove-type old-both`. See the description on page 60 (*Paragraph + document BFF*) of the [Arxiv paper](https://arxiv.org/pdf/2406.11794) for an english description of this method.

As a side note, `--remove-type both` does a slightly more sophisticated thing that is a bit more "true" and accounts for ngrams that span paragraphs for deduplication (whereas `old-both` does not consider ngrams that span paragraphs)

Thanks for pointing this out! I'll update the readme to rectify this mistake!
